compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -415,10 +416,38 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             match obligation.self_ty().skip_binder().kind() {
                 // Fast path to avoid evaluating an obligation that trivally holds.
                 // There may be more bounds, but these are checked by the regular path.
-                ty::FnPtr(..) => continue,
-                ty::Param(..) => {}
-                // No other types can implement `FnPtr`
-                _ => return true,
+                ty::FnPtr(..) => return false,
+                // These may potentially implement `FnPtr`
+                ty::Placeholder(..)
+                | ty::Dynamic(_, _, _)
+                | ty::Alias(_, _)
+                | ty::Infer(_)
+                | ty::Bound(_, _)
```

Comment:
don't have to fix this, but this can be `unreachable!` you cannot have an escaping bound type as the self type